components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,633 @@
+            jwtHeader = addCertToHeader(endUserName);
+        }
+        return jwtHeader;
+    }
+
+    public String buildBody(Map<String, Object> saml2Assertions) throws AppManagementException {
+
+        StringBuilder jwtBuilder = new StringBuilder();
+
+        /* Populate claims from SAML Assertion if "AddClaimsSelectively" property is set to true,
+         else add all claims values available in user profile */
+        if (addClaimsSelectively) {
+            Map<String, String> standardClaims = populateStandardClaims(saml2Assertions);
+            if (standardClaims != null) {
+                jwtBuilder.append(buildJWTBody(standardClaims));
+            }
+        } else {
+            Map<String, String> customClaims = populateCustomClaims(saml2Assertions);
+            if (customClaims != null) {
+                jwtBuilder.append(buildJWTBody(customClaims));
+            }
+        }
+
+        return jwtBuilder.toString();
+    }
+
+    private String buildJWTBody(Map<String, String> claims) {
+
+        StringBuilder jwtBuilder = new StringBuilder();
+        jwtBuilder.append("{");
+        if (claims != null) {
+            Iterator<Map.Entry<String, String>> entryIterator = claims.entrySet().iterator();
+            while (entryIterator.hasNext()) {
+                Map.Entry<String, String> entry = entryIterator.next();
+                String key = entry.getKey();
+                if("exp".equals(key) || "nbf".equals(key) || "iat".equals(key)) {
+                    //These values should be numbers.
+                    jwtBuilder.append("\"" + key + "\":" + entry.getValue() + ",");
+                }
+                else {
+                    jwtBuilder.append("\"" + key + "\":\"" + entry.getValue() + "\",");
+                }
+            }
+        }
+
+        if (jwtBuilder.length() > 1) {
+            jwtBuilder.delete(jwtBuilder.length() - 1, jwtBuilder.length());
+        }
+
+        jwtBuilder.append("}");
+        return jwtBuilder.toString();
+    }
+
+    public abstract Map<String, String> populateStandardClaims(Map<String, Object> saml2Assertions)
+            throws AppManagementException ;
+    public abstract Map<String, String> populateCustomClaims(Map<String, Object> saml2Assertions)
+            throws AppManagementException ;
+    
+
+    public ClaimsRetriever getClaimsRetriever() {
+        return claimsRetriever;
+    }
+
+    /**
+     * Helper method to sign the JWT
+     *
+     * @param assertion
+     * @param endUserName
+     * @return signed assertion
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    private byte[] signJWT(String assertion, String endUserName)
+            throws AppManagementException {
+
+        try {
+            String tenantDomain = MultitenantUtils.getTenantDomain(endUserName);
+            int tenantId = getTenantId(endUserName);
+
+            Key privateKey = null;
+
+            if (!(privateKeys.containsKey(tenantId))) {
```

Comment:
We can make this more elegant, 
1. get from the map, if its is null then load the key
2. no need to concern about race conditions here.